File: dream-fronend/src/app/thongke/thongke.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpHeaders, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';

export interface TopSanPhamResponse {
  tenSanPham: string;  // Tên sản phẩm
  tongSoLuong: number; // Tổng số lượng bán được
}

export interface ThongKeHomNayResponse {
  soHoaDon: number;         // Số hóa đơn trong ngày hôm nay
  tongDoanhThu: number;     // Tổng doanh thu trong ngày hôm nay
  soKhachHang: number;      // Số khách hàng trong ngày hôm nay
}
export interface ThongKeResponse {
  soHoaDon: number;
  tongDoanhThu: number;
  soKhachHang: number;
}

export interface ThongKeThangResponse {
  thang: number; // Có thể là tháng hoặc năm, tùy vào API trả về
  tongDoanhThu: number;
}
export interface ThongKeThangNayResponse {
  ngay: number; // Có thể là tháng hoặc năm, tùy vào API trả về
  tongDoanhThu: number;
}
export interface ThongKeRequest {
  startDate: string | null;
  endDate: string | null;
}

@Injectable({
  providedIn: 'root',
})
export class ThongKeService {
  private apiUrl = 'http://localhost:8080/api/thong-ke';

  constructor(private http: HttpClient) {}

  // Lấy dữ liệu thống kê tổng quan
  thongKeTongQuan(type: string): Observable<ThongKeResponse> {
    return this.http.get<ThongKeResponse>(`${this.apiUrl}/${type}`);
  }

// Lấy dữ liệu thống kê doanh thu từng tháng trong năm được chọn
thongKeTungThangNam(year: number): Observable<ThongKeThangResponse[]> {
  return this.http.get<ThongKeThangResponse[]>(`${this.apiUrl}/nam-nay/thang`, {
    params: { year: year.toString() }
  });
}

  // Lấy dữ liệu thống kê doanh thu theo từng năm
  thongKeTungNam(): Observable<ThongKeThangResponse[]> {
    return this.http.get<ThongKeThangResponse[]>(`${this.apiUrl}/tat-ca/nam`);
  }

  thongKeTungNgayTrongThang(month: number, year: number): Observable<ThongKeThangNayResponse[]> {
    return this.http.get<ThongKeThangNayResponse[]>(
      `${this.apiUrl}/thang-nay/ngay?month=${month}&year=${year}`
    );
  }
  
  // Lấy dữ liệu thống kê doanh thu ngày hôm nay
  thongKeHomNay(): Observable<ThongKeHomNayResponse> {
    return this.http.get<ThongKeHomNayResponse>(`${this.apiUrl}/hom-nay`);
  }

   // Lấy top sản phẩm bán chạy trong ngày hôm nay
   topSanPhamHomNay(page: number, size: number): Observable<TopSanPhamResponse[]> {
    return this.http.get<TopSanPhamResponse[]>(`${this.apiUrl}/hom-nay/top-san-pham?page=${page}&size=${size}`);
  }

  // Lấy top sản phẩm bán chạy theo tháng và năm truyền vào
topSanPhamTheoThangVaNam(thang: number, nam: number, page: number, size: number): Observable<TopSanPhamResponse[]> {
  return this.http.get<TopSanPhamResponse[]>(
    `${this.apiUrl}/top-san-pham?thang=${thang}&nam=${nam}&page=${page}&size=${size}`
  );
}

  // Lấy top sản phẩm bán chạy theo năm được chọn
topSanPhamTheoNam(nam: number, page: number, size: number): Observable<TopSanPhamResponse[]> {
  return this.http.get<TopSanPhamResponse[]>(
    `${this.apiUrl}/top-san-pham-nam?nam=${nam}&page=${page}&size=${size}`
  );
}


  // Lấy top sản phẩm bán chạy tất cả thời gian
  topSanPhamTatCa(page: number, size: number): Observable<TopSanPhamResponse[]> {
    return this.http.get<TopSanPhamResponse[]>(`${this.apiUrl}/tat-ca/top-san-pham?page=${page}&size=${size}`);
  }

   // Phương thức để gọi API lấy tổng quan theo tháng và năm
   getThongKeTongQuan(month: number, year: number): Observable<ThongKeResponse> {
    const params = new HttpParams()
      .set('month', month.toString())
      .set('year', year.toString());

    return this.http.get<ThongKeResponse>(`${this.apiUrl}/thongke/tongquan`, { params });
  }



  // Trong thongke.service.ts
  thongKeTheoKhoangThoiGian(
    startDate: string | null, 
    endDate: string | null
    ): Observable<ThongKeResponse> {
    const headers = new HttpHeaders({
      'Authorization': `Bearer ${localStorage.getItem('access_token')}`
    });
  
    let params = new HttpParams();
    if (startDate) {
      params = params.append('startDate', startDate); // Format: yyyy-MM-dd
    }
    if (endDate) {
      params = params.append('endDate', endDate); // Format: yyyy-MM-dd
    }
  
    return this.http.get<ThongKeResponse>(
      `${this.apiUrl}/khoang-thoi-gian`,
      { params, headers }
    );
    }
  
  
    // Thêm phương thức mới
    topSanPhamTheoKhoangNgay(
      startDate: string | null,
      endDate: string | null,
      page: number,
      size: number
    ): Observable<TopSanPhamResponse[]> {
      const headers = new HttpHeaders({
        'Authorization': `Bearer ${localStorage.getItem('access_token')}`
      });
  
      let params = new HttpParams()
        .append('page', page.toString())
        .append('size', size.toString());
  
      if (startDate) {
        params = params.append('startDate', startDate);
      }
      if (endDate) {
        params = params.append('endDate', endDate);
      }
  
      return this.http.get<TopSanPhamResponse[]>(
        `${this.apiUrl}/khoang-ngay/top-san-pham`,
        { params, headers }
      );
    }
}

File: dream-backend/src/main/java/com/example/dreambackend/services/thongke/ThongKeService.java
package com.example.dreambackend.services.thongke;

import com.example.dreambackend.repositories.HoaDonChiTietRepository;
import com.example.dreambackend.repositories.HoaDonRepository;
import com.example.dreambackend.responses.*;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;

import java.time.DayOfWeek;
import java.time.LocalDate;
import java.time.YearMonth;
import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

@Service
public class ThongKeService {

    @Autowired
    private HoaDonRepository hoaDonRepository;
    @Autowired
    private HoaDonChiTietRepository hoaDonChiTietRepository;

    public ThongKeResponse thongKeTongQuan(String type) {
        LocalDate startDate = null;
        LocalDate endDate = null;

        switch (type) {
            case "Hôm nay":
                startDate = LocalDate.now();
                endDate = LocalDate.now();
                break;
            case "Tất cả":
                // Không thiết lập startDate và endDate
                break;
            default:
                throw new IllegalArgumentException("Invalid type: " + type);
        }

        return hoaDonRepository.getTongQuan(startDate, endDate);
    }

    // Phương thức để lấy tổng quan theo tháng và năm
    public ThongKeResponse getTongQuanTheoThangVaNam(int month, int year) {
        ThongKeResponse response = hoaDonRepository.getTongQuanTheoThangVaNam(month, year);
        if (response == null) {
            return new ThongKeResponse(0, 0, 0);
        }
        return response;
    }

    public List<ThongKeThangResponse> thongKeTungThang(int year) {
        return hoaDonRepository.getDoanhThuTungThangTheoNam(year);
    }

    public List<ThongKeThangResponse> thongKeTungNam() {
        List<Object[]> results = hoaDonRepository.getDoanhThuTungNam();
        return results.stream()
                .filter(obj -> obj[0] != null && obj[1] != null) // lọc bỏ các dòng có dữ liệu null
                .map(obj -> new ThongKeThangResponse((Integer) obj[0], (Double) obj[1]))
                .collect(Collectors.toList());
    }
    // Thống kê doanh thu từng ngày trong tháng
    public List<ThongKeThangNayResponse> thongKeTungNgayTrongThang(int month, int year) {
        return hoaDonRepository.getDoanhThuTheoNgayTheoThang(month, year);
    }

    // Thống kê doanh thu ngày hôm nay
    public ThongKeHomNayResponse thongKeHomNay() {
        return hoaDonRepository.getDoanhThuHomNay();
    }
    // Thống kê sản phẩm bán chạy nhất trong ngày hôm nay
    public List<TopSanPhamResponse> topSanPhamHomNay() {
        return hoaDonChiTietRepository.getTopSanPhamHomNay(PageRequest.of(0, 5)).getContent();
    }

    public List<TopSanPhamResponse> topSanPhamTheoThangVaNam(int thang, int nam) {
        // Xác định ngày bắt đầu và kết thúc của tháng
        LocalDate startDate = LocalDate.of(nam, thang, 1);
        LocalDate endDate = startDate.withDayOfMonth(startDate.lengthOfMonth());

        Page<TopSanPhamResponse> pageResult = hoaDonChiTietRepository
                .getTopSanPhamTheoThangVaNam(PageRequest.of(0, 5), startDate, endDate);

        // Trả về danh sách sản phẩm (dù ít hơn 5 vẫn trả về đầy đủ)
        return pageResult != null ? pageResult.getContent() : new ArrayList<>();
    }



    public List<TopSanPhamResponse> topSanPhamTheoNam(int nam) {
        LocalDate startDate = LocalDate.of(nam, 1, 1);
        LocalDate endDate = LocalDate.of(nam, 12, 31);

        Pageable pageable = PageRequest.of(0, 5);
        Page<TopSanPhamResponse> topSanPhamPage = hoaDonChiTietRepository.getTopSanPhamTheoNam(pageable, startDate, endDate);

        // Kiểm tra xem có ít hơn 5 sản phẩm không và trả về tất cả nếu ít hơn 5
        if (topSanPhamPage.getTotalElements() < 5) {
            return topSanPhamPage.getContent(); // Trả về tất cả sản phẩm có trong năm
        }

        // Nếu có đủ 5 sản phẩm, trả về top 5
        return topSanPhamPage.getContent();
    }


    // Thống kê sản phẩm bán chạy nhất tất cả thời gian
    public List<TopSanPhamResponse> topSanPhamTatCa() {
        return hoaDonChiTietRepository.getTopSanPhamTatCa(PageRequest.of(0, 5)).getContent();
    }
}
